src/nvim/spell.c

```diff
@@ -1309,22 +1310,32 @@ size_t spell_move_to(win_T *wp, int dir, bool allwords, bool curline, hlf_T *att
                             ? p - buf + (ptrdiff_t)len
                             : p - buf) > wp->w_cursor.col)) {
             bool can_spell;
+
+            char *decor_error = NULL;
+
+            col = (colnr_T)(p - buf);
+
+            (void)decor_provider_spell(wp, lnum - 1, col, &decor_error);
+            DecorState tmp_state = { 0 };
+            decor_redraw_start(wp->w_buffer, lnum - 1, &tmp_state);
+            decor_redraw_line(wp->w_buffer, lnum - 1, &tmp_state);
+            decor_redraw_col(wp->w_buffer, col, col, false, &tmp_state);
```

Comment:
instead of faking a redraw like this, which could be quite expensive, why not add a new `on_spell` callback to the provider? then the provider itself can decide what works needs to be done, used cached info from the last redraw, or whatever.

Reply:
This makes sense. 

One note, however. In the context of our current TS highlighter, you need to do pretty everything `on_line` does already, the only waste is the additional marks that get added to the temporary decor_state.

Also only decor_providers with `spell = true` are called so it's not that expensive.

This mean you can emulate an `on_spell` callback with 2 decor providers:

```lua
a.nvim_set_decoration_provider(ns, {
  on_line = on_line,
  spell = false,
})

a.nvim_set_decoration_provider(ns2, {
  on_line = on_spell,
  spell = true,
})
```

So either method works well IMO, it's just a case of which one has a nicer API.